File: go/session/session_test.go
package session

import (
	"bytes"
	"crypto/ecdh"
	srand "crypto/rand"
	"encoding/hex"
	"github.com/jc-lab/jscp/go/cryptoutil"
	"github.com/jc-lab/jscp/go/payloadpb"
	"github.com/jc-lab/jscp/go/sessionstate"
	"github.com/stretchr/testify/assert"
	"golang.org/x/crypto/ed25519"
	"math/rand/v2"
	"testing"
	"time"
)

var ed25519Algorithm = &cryptoutil.Ed25519Algorithm{}
var x25519Algorithm = &cryptoutil.X25519Algorithm{}
var p256DhAlgorithm = &p256DhAlgorithmImpl{}

type p256DhAlgorithmImpl struct{}

func (p *p256DhAlgorithmImpl) GetType() payloadpb.DHAlgorithm {
	return payloadpb.DHAlgorithm_DHECC
}

func (p *p256DhAlgorithmImpl) Generate() (*cryptoutil.DHKeyPair, error) {
	curve := ecdh.P256()
	privateKey, err := curve.GenerateKey(srand.Reader)
	if err != nil {
		return nil, err
	}
	eccPrivateKey := cryptoutil.NewECCPrivateKey(nil, privateKey, cryptoutil.KeyUsageDH)

	return &cryptoutil.DHKeyPair{
		Private: eccPrivateKey,
		Public:  eccPrivateKey.GetDHPublic(),
	}, nil
}

func (p *p256DhAlgorithmImpl) UnmarshalDHPublicKey(input []byte) (cryptoutil.DHPublicKey, error) {
	//TODO implement me
	panic("implement me")
}

var testCases = []struct {
	name                 string
	serverAdditional     []byte
	clientAdditional     []byte
	expectAdditional     bool
	serverUseSignature   bool
	clientUseSignature   bool
	serverUseDHSignature string
	clientUseDHSignature string
}{
	{
		name:             "no_static-no_static-wo_additional",
		serverAdditional: nil,
		clientAdditional: nil,
		expectAdditional: false,
	},
	{
		name:             "no_static-no_static-with_additional",
		serverAdditional: []byte("i am server"),
		clientAdditional: []byte("i am client"),
		expectAdditional: true,
	},

	{
		name:               "static(signature)-static(signature)-wo_additional",
		serverAdditional:   nil,
		clientAdditional:   nil,
		expectAdditional:   false,
		serverUseSignature: true,
		clientUseSignature: true,
	},
	{
		name:               "static(signature)-static(signature)-with_additional",
		serverAdditional:   []byte("i am server"),
		clientAdditional:   []byte("i am client"),
		expectAdditional:   true,
		serverUseSignature: true,
		clientUseSignature: true,
	},

	{
		name:                 "static(x25519)-static(x25519)-wo_additional",
		serverAdditional:     nil,
		clientAdditional:     nil,
		expectAdditional:     false,
		serverUseSignature:   true,
		clientUseSignature:   true,
		serverUseDHSignature: "x25519",
		clientUseDHSignature: "x25519",
	},
	{
		name:                 "static(x25519)-static(x25519)-with_additional",
		serverAdditional:     []byte("i am server"),
		clientAdditional:     []byte("i am client"),
		expectAdditional:     true,
		serverUseSignature:   true,
		clientUseSignature:   true,
		serverUseDHSignature: "x25519",
		clientUseDHSignature: "x25519",
	},

	{
		name:                 "static(p256)-static(p256)-wo_additional",
		serverAdditional:     nil,
		clientAdditional:     nil,
		expectAdditional:     false,
		serverUseSignature:   true,
		clientUseSignature:   true,
		serverUseDHSignature: "p256",
		clientUseDHSignature: "p256",
	},
	{
		name:                 "static(p256)-static(p256)-with_additional",
		serverAdditional:     []byte("i am server"),
		clientAdditional:     []byte("i am client"),
		expectAdditional:     true,
		serverUseSignature:   true,
		clientUseSignature:   true,
		serverUseDHSignature: "p256",
		clientUseDHSignature: "p256",
	},
	{
		name:                 "no_static(p256)-static(p256)-with_additional",
		serverAdditional:     []byte("i am server"),
		clientAdditional:     []byte("i am client"),
		expectAdditional:     true,
		serverUseSignature:   false,
		clientUseSignature:   true,
		serverUseDHSignature: "",
		clientUseDHSignature: "p256",
	},
	{
		name:                 "static(p256)-no_static(p256)-with_additional",
		serverAdditional:     []byte("i am server"),
		clientAdditional:     []byte("i am client"),
		expectAdditional:     true,
		serverUseSignature:   true,
		clientUseSignature:   false,
		serverUseDHSignature: "p256",
		clientUseDHSignature: "",
	},

	{
		name:                 "data communication",
		serverAdditional:     []byte("i am server"),
		clientAdditional:     []byte("i am client"),
		expectAdditional:     true,
		serverUseSignature:   true,
		clientUseSignature:   true,
		serverUseDHSignature: "p256",
		clientUseDHSignature: "p256",
	},
}

func TestCommunicationDynamic(t *testing.T) {
	var seed [32]byte
	srand.Reader.Read(seed[:])
	randReader := rand.NewChaCha8(seed)
	buf := make([]byte, 65536)

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			var serverPrivateKey cryptoutil.PrivateKey
			var clientPrivateKey cryptoutil.PrivateKey

			if tc.serverUseSignature {
				if tc.serverUseDHSignature != "" {
					var algo cryptoutil.DHAlgorithm
					switch tc.serverUseDHSignature {
					case "x25519":
						algo = x25519Algorithm
					case "p256":
						algo = p256DhAlgorithm
					}
					if keyPair, err := algo.Generate(); err != nil {
						t.Fatal(err)
					} else {
						serverPrivateKey = keyPair.Private
					}
				} else {
					var err error
					if serverPrivateKey, err = ed25519Algorithm.Generate(); err != nil {
						t.Fatal(err)
					}
				}
			}

			if tc.clientUseSignature {
				if tc.clientUseDHSignature != "" {
					var algo cryptoutil.DHAlgorithm
					switch tc.clientUseDHSignature {
					case "x25519":
						algo = x25519Algorithm
					case "p256":
						algo = p256DhAlgorithm
					}
					if keyPair, err := algo.Generate(); err != nil {
						t.Fatal(err)
					} else {
						clientPrivateKey = keyPair.Private
					}
				} else {
					var err error
					if clientPrivateKey, err = ed25519Algorithm.Generate(); err != nil {
						t.Fatal(err)
					}
				}
			}

			// Create channels for communication between client and server
			clientToServer := make(chan *payloadpb.Payload, 1)
			serverToClient := make(chan *payloadpb.Payload, 1)

			// Create server and client with their respective send functions
			server, err := NewSession(false, func(payload *payloadpb.Payload) {
				serverToClient <- payload
			}, serverPrivateKey)
			if err != nil {
				t.Fatal(err)
			}

			client, err := NewSession(true, func(payload *payloadpb.Payload) {
				clientToServer <- payload
			}, clientPrivateKey)
			if err != nil {
				t.Fatal(err)
			}

			client.symmetricStateFactory = func() sessionstate.SymmetricState {
				return &testSymmetricState{}
			}
			server.symmetricStateFactory = client.symmetricStateFactory
			client.localState = &testSymmetricState{}
			client.remoteState = &testSymmetricState{}
			server.localState = &testSymmetricState{}
			server.remoteState = &testSymmetricState{}

			// Start goroutines to handle message passing
			go func() {
				for payload := range serverToClient {
					err := client.HandleReceive(payload)
					if err != nil {
						t.Fatalf("client.HandleReceive error: %v", err)
					}
				}
			}()

			go func() {
				for payload := range clientToServer {
					err := server.HandleReceive(payload)
					if err != nil {
						t.Fatalf("server.HandleReceive error: %v", err)
					}
				}
			}()

			// Start handshakes
			clientHandshakeCh, err := client.Handshake(tc.clientAdditional)
			if err != nil {
				t.Fatalf("client.Handshake error: %v", err)
			}

			serverHandshakeCh, err := server.Handshake(tc.serverAdditional)
			if err != nil {
				t.Fatalf("server.Handshake error: %v", err)
			}

			// Wait for both handshakes to complete with timeout
			timer := time.NewTimer(500 * time.Millisecond)
			defer timer.Stop()

			var clientResult, serverResult *HandshakeResult

			// Wait for client result
			select {
			case result := <-clientHandshakeCh:
				if result.Error != nil {
					t.Fatalf("client handshake failed: %v", result.Error)
				}
				clientResult = result
			case <-timer.C:
				t.Fatal("timeout waiting for client handshake")
			}

			// Wait for server result
			select {
			case result := <-serverHandshakeCh:
				if result.Error != nil {
					t.Fatalf("server handshake failed: %v", result.Error)
				}
				serverResult = result
			case <-timer.C:
				t.Fatal("timeout waiting for server handshake")
			}

			// Verify additional data if expected
			if tc.expectAdditional {
				if clientResult.PeerAdditionalData == nil {
					t.Error("client result: expected peer additional data, got nil")
				} else if !bytes.Equal(clientResult.PeerAdditionalData, tc.serverAdditional) {
					t.Errorf("client result: peer additional data mismatch\nexpected: %v\ngot: %v",
						tc.serverAdditional, clientResult.PeerAdditionalData)
				}

				if serverResult.PeerAdditionalData == nil {
					t.Error("server result: expected peer additional data, got nil")
				} else if !bytes.Equal(serverResult.PeerAdditionalData, tc.clientAdditional) {
					t.Errorf("server result: peer additional data mismatch\nexpected: %v\ngot: %v",
						tc.clientAdditional, serverResult.PeerAdditionalData)
				}
			} else {
				if clientResult.PeerAdditionalData != nil {
					t.Error("client result: expected nil peer additional data")
				}
				if serverResult.PeerAdditionalData != nil {
					t.Error("server result: expected nil peer additional data")
				}
			}

			if tc.serverUseSignature {
				compareSignaturePublicKey(t, serverPrivateKey.GetPublic(), clientResult.RemotePublicKey)
			}
			if tc.clientUseSignature {
				compareSignaturePublicKey(t, clientPrivateKey.GetPublic(), serverResult.RemotePublicKey)
			}

			clientStateA := client.localState.(*testSymmetricState)
			clientStateB := server.remoteState.(*testSymmetricState)
			assert.Equal(t, clientStateA.recordedKeys, clientStateB.recordedKeys)

			serverStateA := client.remoteState.(*testSymmetricState)
			serverStateB := server.localState.(*testSymmetricState)
			assert.Equal(t, serverStateA.recordedKeys, serverStateB.recordedKeys)

			assert.Equal(t, clientStateA.recordedKeys, serverStateB.recordedKeys)

			index := 0

			if tc.serverUseDHSignature != "" {
				out, _ := client.ephemeralKeyPair.Private.DH(server.staticKeyPair.GetPublic().(cryptoutil.DHPublicKey))
				assert.Equal(t, out, clientStateA.recordedKeys[index])
				index++
			}
			if tc.clientUseDHSignature != "" {
				out, _ := client.staticKeyPair.(cryptoutil.DHPrivateKey).DH(server.ephemeralKeyPair.Public)
				assert.Equal(t, out, clientStateA.recordedKeys[index])
				index++
			}
			if true {
				out, _ := client.ephemeralKeyPair.Private.DH(server.ephemeralKeyPair.Public)
				assert.Equal(t, out, clientStateA.recordedKeys[index])
				index++
			}
			assert.Equal(t, len(clientStateA.recordedKeys), index)

			for i := 0; i < 100; i++ {
				randReader.Read(buf)
				server.Write(buf)
				recv, err := client.Read()
				assert.NoError(t, err)
				assert.Equal(t, buf, recv)

				randReader.Read(buf)
				client.Write(buf)
				recv, err = server.Read()
				assert.NoError(t, err)
				assert.Equal(t, buf, recv)
			}

			// Close channels
			close(clientToServer)
			close(serverToClient)
		})
	}
}

func TestCommunicationDeterministic(t *testing.T) {
	for _, tc := range testCases {
		if tc.clientUseDHSignature == "p256" || tc.serverUseDHSignature == "p256" {
			continue
		}
		t.Run(tc.name, func(t *testing.T) {
			var serverPrivateKey cryptoutil.PrivateKey
			var clientPrivateKey cryptoutil.PrivateKey

			if tc.serverUseSignature {
				if tc.serverUseDHSignature != "" {
					serverPrivateKey = decodeX25519Key(t, "27c0beabad79e248d280c2cb48f4704d61904a8447e1b5fb95e914432e772d6d")
				} else {
					serverPrivateKey = decodeEd25519Key(t, "adde2498a77bcded1a89416093bf071fd5f5d89e6caa2e8c4baff1b837bb6e1897d93a4c69d3ab58ff63c341e5564d07cbf9f4330f64ad87ab37c48ce4c7732d")
				}
			}

			if tc.clientUseSignature {
				if tc.clientUseDHSignature != "" {
					clientPrivateKey = decodeX25519Key(t, "2e93f3dc9eef2ee59282b13aa90a8aba079cf5d584959795d72c2bfd8c9d9555")
				} else {
					clientPrivateKey = decodeEd25519Key(t, "6f3ae78202e786109b77afeb6a8ebcd7dbbf31b534f755582ea3703a604610c82c0c79fb73469f3b71badb9a6d995ab2dcbb991db2305e72906ac2640255d03a")
				}
			}

			// Create channels for communication between client and server
			clientToServer := make(chan *payloadpb.Payload, 1)
			serverToClient := make(chan *payloadpb.Payload, 1)

			// Create server and client with their respective send functions
			server, err := NewSession(false, func(payload *payloadpb.Payload) {
				serverToClient <- payload
			}, serverPrivateKey)
			if err != nil {
				t.Fatal(err)
			}

			client, err := NewSession(true, func(payload *payloadpb.Payload) {
				clientToServer <- payload
			}, clientPrivateKey)
			if err != nil {
				t.Fatal(err)
			}

			client.symmetricStateFactory = func() sessionstate.SymmetricState {
				return &testSymmetricState{}
			}
			server.symmetricStateFactory = client.symmetricStateFactory
			client.localState = &testSymmetricState{}
			client.remoteState = &testSymmetricState{}
			server.localState = &testSymmetricState{}
			server.remoteState = &testSymmetricState{}

			// Start goroutines to handle message passing
			go func() {
				for payload := range serverToClient {
					err := client.HandleReceive(payload)
					if err != nil {
						t.Fatalf("client.HandleReceive error: %v", err)
					}
				}
			}()

			go func() {
				for payload := range clientToServer {
					err := server.HandleReceive(payload)
					if err != nil {
						t.Fatalf("server.HandleReceive error: %v", err)
					}
				}
			}()

			// Start handshakes
			clientHandshakeCh, err := client.Handshake(tc.clientAdditional)
			if err != nil {
				t.Fatalf("client.Handshake error: %v", err)
			}

			serverHandshakeCh, err := server.Handshake(tc.serverAdditional)
			if err != nil {
				t.Fatalf("server.Handshake error: %v", err)
			}

			// Wait for both handshakes to complete with timeout
			timer := time.NewTimer(500 * time.Millisecond)
			defer timer.Stop()

			var clientResult, serverResult *HandshakeResult

			// Wait for client result
			select {
			case result := <-clientHandshakeCh:
				if result.Error != nil {
					t.Fatalf("client handshake failed: %v", result.Error)
				}
				clientResult = result
			case <-timer.C:
				t.Fatal("timeout waiting for client handshake")
			}

			// Wait for server result
			select {
			case result := <-serverHandshakeCh:
				if result.Error != nil {
					t.Fatalf("server handshake failed: %v", result.Error)
				}
				serverResult = result
			case <-timer.C:
				t.Fatal("timeout waiting for server handshake")
			}

			// Verify additional data if expected
			if tc.expectAdditional {
				if clientResult.PeerAdditionalData == nil {
					t.Error("client result: expected peer additional data, got nil")
				} else if !bytes.Equal(clientResult.PeerAdditionalData, tc.serverAdditional) {
					t.Errorf("client result: peer additional data mismatch\nexpected: %v\ngot: %v",
						tc.serverAdditional, clientResult.PeerAdditionalData)
				}

				if serverResult.PeerAdditionalData == nil {
					t.Error("server result: expected peer additional data, got nil")
				} else if !bytes.Equal(serverResult.PeerAdditionalData, tc.clientAdditional) {
					t.Errorf("server result: peer additional data mismatch\nexpected: %v\ngot: %v",
						tc.clientAdditional, serverResult.PeerAdditionalData)
				}
			} else {
				if clientResult.PeerAdditionalData != nil {
					t.Error("client result: expected nil peer additional data")
				}
				if serverResult.PeerAdditionalData != nil {
					t.Error("server result: expected nil peer additional data")
				}
			}

			if tc.serverUseSignature {
				compareSignaturePublicKey(t, serverPrivateKey.GetPublic(), clientResult.RemotePublicKey)
			}
			if tc.clientUseSignature {
				compareSignaturePublicKey(t, clientPrivateKey.GetPublic(), serverResult.RemotePublicKey)
			}

			clientStateA := client.localState.(*testSymmetricState)
			clientStateB := server.remoteState.(*testSymmetricState)
			assert.Equal(t, clientStateA.recordedKeys, clientStateB.recordedKeys)

			serverStateA := client.remoteState.(*testSymmetricState)
			serverStateB := server.localState.(*testSymmetricState)
			assert.Equal(t, serverStateA.recordedKeys, serverStateB.recordedKeys)

			assert.Equal(t, clientStateA.recordedKeys, serverStateB.recordedKeys)

			index := 0

			if tc.serverUseDHSignature != "" {
				out, _ := client.ephemeralKeyPair.Private.DH(server.staticKeyPair.GetPublic().(cryptoutil.DHPublicKey))
				assert.Equal(t, out, clientStateA.recordedKeys[index])
				index++
			}
			if tc.clientUseDHSignature != "" {
				out, _ := client.staticKeyPair.(cryptoutil.DHPrivateKey).DH(server.ephemeralKeyPair.Public)
				assert.Equal(t, out, clientStateA.recordedKeys[index])
				index++
			}
			if true {
				out, _ := client.ephemeralKeyPair.Private.DH(server.ephemeralKeyPair.Public)
				assert.Equal(t, out, clientStateA.recordedKeys[index])
				index++
			}
			assert.Equal(t, len(clientStateA.recordedKeys), index)

			// Close channels
			close(clientToServer)
			close(serverToClient)
		})
	}
}

func compareSignaturePublicKey(t *testing.T, expected cryptoutil.PublicKey, actual cryptoutil.PublicKey) {
	aProto, err := expected.MarshalToProto()
	if err != nil {
		t.Fatal(err)
	}
	assert.NotNil(t, actual)
	if actual == nil {
		return
	}
	bProto, err := actual.MarshalToProto()
	if err != nil {
		t.Fatal(err)
	}
	assert.Equal(t, aProto, bProto)
}

type testSymmetricState struct {
	sessionstate.SymmetricStateImpl
	recordedKeys [][]byte
}

func (t *testSymmetricState) MixKey(cipher cryptoutil.CipherAlgorithm, key []byte) {
	t.recordedKeys = append(t.recordedKeys, key)
	t.SymmetricStateImpl.MixKey(cipher, key)
}

func decodeX25519Key(t *testing.T, input string) *cryptoutil.X25519PrivateKey {
	raw, err := hex.DecodeString(input)
	if err != nil {
		t.Fatal(err)
	}
	c := ecdh.X25519()
	priv, err := c.NewPrivateKey(raw)
	if err != nil {
		t.Fatal(err)
	}
	return &cryptoutil.X25519PrivateKey{
		Key: priv,
	}
}

func decodeEd25519Key(t *testing.T, input string) *cryptoutil.Ed25519PrivateKey {
	raw, err := hex.DecodeString(input)
	if err != nil {
		t.Fatal(err)
	}
	priv := ed25519.PrivateKey(raw)
	return &cryptoutil.Ed25519PrivateKey{
		Key: priv,
	}
}

File: java/src/test/kotlin/kr/jclab/jscp/SessionTest.kt
package kr.jclab.jscp

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.future.await
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withTimeout
import kr.jclab.jscp.crypto.*
import kr.jclab.jscp.payload.JcspPayload
import kr.jclab.jscp.state.SymmetricState
import org.assertj.core.api.Assertions.assertThat
import org.bouncycastle.util.encoders.Hex
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import java.security.SecureRandom

class SessionTest {
    data class TestCase(
        val name: String,
        val serverAdditional: ByteArray?,
        val clientAdditional: ByteArray?,
        val expectAdditional: Boolean,
        val serverUseSignature: Boolean,
        val clientUseSignature: Boolean,
        val serverUseDHSignature: String?,
        val clientUseDHSignature: String?
    )

    private val testCases = listOf(
        TestCase(
            name = "no_static-no_static-wo_additional",
            serverAdditional = null,
            clientAdditional = null,
            expectAdditional = false,
            serverUseSignature = false,
            clientUseSignature = false,
            serverUseDHSignature = null,
            clientUseDHSignature = null
        ),
        TestCase(
            name = "no_static-no_static-with_additional",
            serverAdditional = "i am server".toByteArray(),
            clientAdditional = "i am client".toByteArray(),
            expectAdditional = true,
            serverUseSignature = false,
            clientUseSignature = false,
            serverUseDHSignature = null,
            clientUseDHSignature = null
        ),
        TestCase(
            name = "static(signature)-static(signature)-wo_additional",
            serverAdditional = null,
            clientAdditional = null,
            expectAdditional = false,
            serverUseSignature = true,
            clientUseSignature = true,
            serverUseDHSignature = null,
            clientUseDHSignature = null
        ),
        TestCase(
            name = "static(signature)-static(signature)-with_additional",
            serverAdditional = "i am server".toByteArray(),
            clientAdditional = "i am client".toByteArray(),
            expectAdditional = true,
            serverUseSignature = true,
            clientUseSignature = true,
            serverUseDHSignature = null,
            clientUseDHSignature = null
        ),
        TestCase(
            name = "static(x25519)-static(x25519)-wo_additional",
            serverAdditional = null,
            clientAdditional = null,
            expectAdditional = false,
            serverUseSignature = true,
            clientUseSignature = true,
            serverUseDHSignature = "x25519",
            clientUseDHSignature = "x25519"
        ),
        TestCase(
            name = "static(x25519)-static(x25519)-with_additional",
            serverAdditional = "i am server".toByteArray(),
            clientAdditional = "i am client".toByteArray(),
            expectAdditional = true,
            serverUseSignature = true,
            clientUseSignature = true,
            serverUseDHSignature = "x25519",
            clientUseDHSignature = "x25519"
        ),
        TestCase(
            name = "static(p256)-static(p256)-wo_additional",
            serverAdditional = null,
            clientAdditional = null,
            expectAdditional = false,
            serverUseSignature = true,
            clientUseSignature = true,
            serverUseDHSignature = "p256",
            clientUseDHSignature = "p256"
        ),
        TestCase(
            name = "static(p256)-static(p256)-with_additional",
            serverAdditional = "i am server".toByteArray(),
            clientAdditional = "i am client".toByteArray(),
            expectAdditional = true,
            serverUseSignature = true,
            clientUseSignature = true,
            serverUseDHSignature = "p256",
            clientUseDHSignature = "p256"
        ),
        TestCase(
            name = "no_static(p256)-static(p256)-with_additional",
            serverAdditional = "i am server".toByteArray(),
            clientAdditional = "i am client".toByteArray(),
            expectAdditional = true,
            serverUseSignature = false,
            clientUseSignature = true,
            serverUseDHSignature = "",
            clientUseDHSignature = "p256"
        ),
        TestCase(
            name = "static(p256)-no_static(p256)-with_additional",
            serverAdditional = "i am server".toByteArray(),
            clientAdditional = "i am client".toByteArray(),
            expectAdditional = true,
            serverUseSignature = true,
            clientUseSignature = false,
            serverUseDHSignature = "p256",
            clientUseDHSignature = ""
        ),
        TestCase(
            name = "data communication",
            serverAdditional = "i am server".toByteArray(),
            clientAdditional = "i am client".toByteArray(),
            expectAdditional = true,
            serverUseSignature = true,
            clientUseSignature = true,
            serverUseDHSignature = "p256",
            clientUseDHSignature = "p256"
        )
    )

    @TestFactory
    fun testCommunicationDynamic() = testCases.map { tc ->
        DynamicTest.dynamicTest(tc.name) {
            runTest {
                val secureRandom = SecureRandom()
                val buffer = ByteArray(65536)

                var serverPrivateKey: JscpPrivateKey? = null
                var clientPrivateKey: JscpPrivateKey? = null

                // 서버 키 설정
                if (tc.serverUseSignature) {
                    serverPrivateKey = when (tc.serverUseDHSignature) {
                        "x25519" -> X25519Algorithm.generate().private
                        "p256" -> ECCAlgorithm(KeyUsage.DH).generateKeyPair("P-256").first
                        else -> Ed25519Algorithm.generate().first
                    }
                }

                // 클라이언트 키 설정
                if (tc.clientUseSignature) {
                    clientPrivateKey = when (tc.clientUseDHSignature) {
                        "x25519" -> X25519Algorithm.generate().private
                        "p256" -> ECCAlgorithm(KeyUsage.DH).generateKeyPair("P-256").first
                        else -> Ed25519Algorithm.generate().first
                    }
                }

                // 통신 채널 설정
                val clientToServer = Channel<JcspPayload.Payload>(Channel.UNLIMITED)
                val serverToClient = Channel<JcspPayload.Payload>(Channel.UNLIMITED)
                val clientReceive = Channel<ByteArray>(Channel.UNLIMITED)
                val serverReceive = Channel<ByteArray>(Channel.UNLIMITED)

                // 세션 생성
                val server = Session(
                    initiator = false,
                    staticKeyPair = serverPrivateKey
                ) { payload -> runBlocking { serverToClient.send(payload) } }
                server.onReceive = { data ->
                    runBlocking { serverReceive.send(data) }
                }

                val client = Session(
                    initiator = true,
                    staticKeyPair = clientPrivateKey
                ) { payload -> runBlocking { clientToServer.send(payload) } }
                client.onReceive = { data ->
                    runBlocking { clientReceive.send(data) }
                }

                // 테스트용 SymmetricState 설정
                client.symmetricStateFactory = { TestSymmetricState("client") }
                server.symmetricStateFactory = { TestSymmetricState("server") }

                // 메시지 핸들링
                launch {
                    for (payload in serverToClient) {
                        client.handleReceive(payload)
                    }
                }

                launch {
                    for (payload in clientToServer) {
                        server.handleReceive(payload)
                    }
                }

                // 핸드쉐이크 수행
                val clientHandshakeResult = client.handshake(tc.clientAdditional)
                val serverHandshakeResult = server.handshake(tc.serverAdditional)

                // 결과 검증
                withTimeout(500) {
                    val clientResult = clientHandshakeResult.await()
                    val serverResult = serverHandshakeResult.await()

                    println(clientResult)
                    println(serverResult)

                    // Additional 데이터 검증
                    if (tc.expectAdditional) {
                        assertThat(clientResult.peerAdditionalData)
                            .isEqualTo(tc.serverAdditional)
                        assertThat(serverResult.peerAdditionalData)
                            .isEqualTo(tc.clientAdditional)
                    } else {
                        assertThat(clientResult.peerAdditionalData).isNull()
                        assertThat(serverResult.peerAdditionalData).isNull()
                    }

                    // 공개키 검증
                    if (tc.serverUseSignature) {
                        compareSignaturePublicKey(
                            serverPrivateKey,
                            clientResult.remotePublicKey
                        )
                    }
                    if (tc.clientUseSignature) {
                        compareSignaturePublicKey(
                            clientPrivateKey,
                            serverResult.remotePublicKey
                        )
                    }

                    // SymmetricState 검증
                    val clientStateA = client.localState as TestSymmetricState
                    val clientStateB = server.remoteState as TestSymmetricState
                    assertThat(clientStateA.recordedKeys).hasSize(clientStateB.recordedKeys.size)
                    for (i in 0 until clientStateA.recordedKeys.size) {
                        assertArrayEquals(clientStateA.recordedKeys[i], clientStateB.recordedKeys[i])
                    }

                    val serverStateA = client.remoteState as TestSymmetricState
                    val serverStateB = server.localState as TestSymmetricState
                    assertThat(serverStateA.recordedKeys).hasSize(serverStateB.recordedKeys.size)
                    for (i in 0 until clientStateA.recordedKeys.size) {
                        assertArrayEquals(serverStateA.recordedKeys[i], serverStateB.recordedKeys[i])
                    }

                    // 데이터 통신 테스트
                    repeat(100) {
                        secureRandom.nextBytes(buffer)

                        server.write(buffer)
                        val clientReceived = clientReceive.receive()
                        assertThat(clientReceived).isEqualTo(buffer)

                        secureRandom.nextBytes(buffer)
                        client.write(buffer)
                        val serverReceived = serverReceive.receive()
                        assertThat(serverReceived).isEqualTo(buffer)
                    }
                }

                // 정리
                clientToServer.close()
                serverToClient.close()
            }
        }
    }

    private fun compareSignaturePublicKey(expectedPrivateKey: JscpPrivateKey?, actual: JscpPublicKey?) {
        assertThat(actual?.marshalToProto()).isEqualTo(expectedPrivateKey?.getPublic()?.marshalToProto())
    }
}

class TestSymmetricState(
    val name: String,
) : SymmetricState() {
    val recordedKeys = mutableListOf<ByteArray>()

    override fun mixKey(cipher: CipherAlgorithm, key: ByteArray) {
//        println("[${name}] MIX KEY: ${Hex.toHexString(key)}")
        recordedKeys.add(key)
        super.mixKey(cipher, key)
    }

    override fun mixHash(data: ByteArray) {
//        println("[${name}] MIX HASH: ${Hex.toHexString(data)}")
        super.mixHash(data)
    }
}